core_data_modules/traced_data/io.py

```diff
@@ -439,6 +445,305 @@ def import_coda_to_traced_data_iterable_as_matrix(cls, user, data, key_of_raw, c
+        TracedData objects with the same message id must have the same labels applied, otherwise this exporter will fail.
+
+        :param data: Data to export to Coda V2.
+        :type data: iterable of TracedData
+        :param raw_key: Key in TracedData objects of the raw messages.
+        :type raw_key: str
+        :param creation_date_time_key: Key in TracedData objects of when the message was created.
+        :type creation_date_time_key: str
+        :param message_id_key: Key in TracedData objects of the message id.
+                               Message Ids may be set using TracedDataCoda2IO.add_message_ids.
+        :type message_id_key: str
+        :param scheme_keys: Dictionary of (key in TracedData objects of coded data to export) ->
+                            (Scheme for that key)
+        :type scheme_keys: dict of str -> Scheme
+        :param f: File to write exported JSON file to.
+        :type f: file-like
+        """
+        # Filter data for elements which contain the given raw key
+        data = [td for td in data if raw_key in td]
+
+        cls._assert_uniquely_coded(data, message_id_key, scheme_keys.keys())
+        data = cls._filter_duplicates(data, message_id_key, creation_date_time_key)
+        data = cls._exclude_missing(data, scheme_keys)
+
+        coda_messages = []  # List of Coda V2 Message objects to be exported
+        for td in data:
+            # Export labels for this row which are not Codes.NOT_CODED
+            labels = []
+            for coded_key, scheme in scheme_keys.items():
+                if coded_key in td and scheme.get_code_with_id(td[coded_key]["CodeID"]).control_code != Codes.NOT_CODED:
+                    labels.append(Label.from_firebase_map(td[coded_key]))
+
+            # Create a Coda message object for this row
+            message = Message(
+                message_id=td[message_id_key],
+                text=td[raw_key],
+                creation_date_time_utc=isoparse(td[creation_date_time_key]).astimezone(pytz.utc).isoformat(),
+                labels=labels
+            )
+
+            coda_messages.append(message)
+
+        json.dump([m.to_firebase_map() for m in coda_messages], f, sort_keys=True, indent=2, separators=(", ", ": "))
+
+    @staticmethod
+    def _make_empty_file():
+        """
+        :return: Object which looks like a Coda 2 messages file containing no messages.
+        :rtype: StringIO
+        """
+        return io.StringIO("[]")
+
+    @staticmethod
+    def _dataset_lut_from_messages_file(f):
+        """
+        Creates a lookup table of MessageID -> SchemeID -> Labels from the given Coda 2 messages file.
+
+        :param f: Coda 2 messages file.
+        :type f: file-like
+        :return: Lookup table.
+        :rtype: dict of str -> (dict of str -> list of dict)
+        """
+        coda_dataset = dict()  # of MessageID -> (dict of SchemeID -> list of Label)
+
+        for msg in json.load(f):
```

Comment:
Is there a reason this isn't can't the Schemes etc. objects from data_models?

Reply:
This was first written before those objects were written. Will address in a future PR, once the Label.confidence validation issue is resolved